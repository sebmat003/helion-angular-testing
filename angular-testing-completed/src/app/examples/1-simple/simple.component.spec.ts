import { SimpleComponent } from './simple.component';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { RouterOutlet } from '@angular/router';
import { mockedBooks } from './testing/book.mocks';
import { Category } from './models/book.models';

describe('SimpleComponent', () => {
  let spectator: Spectator<SimpleComponent>;
  let component: SimpleComponent;
  const createComponent = createComponentFactory({
    component: SimpleComponent,
    imports: [RouterOutlet],
    declareComponent: false,
    detectChanges: false,
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('should match the snapshot', () => {
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should display all the books correctly', () => {
    component.filteredBooks = mockedBooks;

    spectator.detectChanges();

    expect(spectator.queryAll('.book')).toHaveLength(mockedBooks.length);
    expect(spectator.queryLast('.book div')?.textContent).toBe(
      'Test 3 - fantasy'
    );
    expect(spectator.query('.book .bestseller')).toBeVisible();
  });

  it('should display a message when no books are available', () => {
    component.filteredBooks = [];

    spectator.detectChanges();

    expect(spectator.query('p')?.textContent).toBe(
      'No available books on the list'
    );
  });

  it.each([
    [
      Category.FANTASY,
      [
        {
          id: 2,
          bestseller: false,
          category: Category.FANTASY,
          name: 'Test 2',
          price: 100,
        },
        {
          id: 3,
          bestseller: false,
          category: Category.FANTASY,
          name: 'Test 3',
          price: 15,
        },
      ],
    ],
    [
      Category.THRILLER,
      [
        {
          id: 1,
          bestseller: true,
          category: Category.THRILLER,
          name: 'Test 1',
          price: 0,
        },
      ],
    ],
  ])(
    'should display filtered list of books for %s category',
    (category, books) => {
      component.filteredBooks = mockedBooks;

      spectator.click(
        spectator
          .queryAll('.categories button')
          .filter((button) => button.textContent === category)[0]
      );
      spectator.detectChanges();

      expect(component.filteredBooks === books);
    }
  );

  describe('the book price', () => {
    beforeEach(() => {
      component.filteredBooks = mockedBooks;

      spectator.detectChanges();
    });

    it('should have a red color when price is equal/below 10', () => {
      expect(spectator.query('.book .price')?.classList['1']).toBe('cheap');
    });

    it('should have a black color when price is above 10', () => {
      expect(spectator.queryLast('.book .price')?.classList.length).toBe(1);
    });
  });
});
