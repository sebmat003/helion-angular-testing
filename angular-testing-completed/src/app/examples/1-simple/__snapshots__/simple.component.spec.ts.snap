// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleComponent should match the snapshot 1`] = `
<app-simple
  activeCategory={[Function Function]}
  books={[Function Array]}
  categories={[Function Array]}
  filteredBooks={[Function Array]}
>
  <div
    class="categories"
  >
    <button
      class="active"
    >
       All 
    </button>
    <button
      data-cy="book-category-button"
    >
       Fantasy 
    </button>
    <button
      data-cy="book-category-button"
    >
       Science Fiction 
    </button>
    <button
      data-cy="book-category-button"
    >
       Biography 
    </button>
    <button
      data-cy="book-category-button"
    >
       Thriller 
    </button>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        My biography - biography
      </p>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       20$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Harry Potter and the Philosopher\`s Stone - fantasy
      </p>
      <span
        class="bestseller"
        data-cy="book-bestseller"
      >
        bestseller!
      </span>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       20$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Harry Potter and the Chamber of Secrets - fantasy
      </p>
      <span
        class="bestseller"
        data-cy="book-bestseller"
      >
        bestseller!
      </span>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       30$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Harry Potter and the Prisoner of Azkaban - fantasy
      </p>
      <span
        class="bestseller"
        data-cy="book-bestseller"
      >
        bestseller!
      </span>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       30$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Harry Potter and the Goblet of Fire - fantasy
      </p>
      <span
        class="bestseller"
        data-cy="book-bestseller"
      >
        bestseller!
      </span>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       30$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Harry Potter and the Order of the Phoenix - fantasy
      </p>
      <span
        class="bestseller"
        data-cy="book-bestseller"
      >
        bestseller!
      </span>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       30$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Day zero - science fiction
      </p>
    </div>
    <p
      class="price cheap"
      data-cy="book-price"
    >
       10$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        My biography - thriller
      </p>
    </div>
    <p
      class="price cheap"
      data-cy="book-price"
    >
       5$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Solaris - science fiction
      </p>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       20$ 
    </p>
  </div><div
    class="book"
    data-cy="book"
  >
    <div>
      <p>
        Ender\`s game - science fiction
      </p>
      <span
        class="bestseller"
        data-cy="book-bestseller"
      >
        bestseller!
      </span>
    </div>
    <p
      class="price"
      data-cy="book-price"
    >
       40$ 
    </p>
  </div>
</app-simple>
`;
